src/transformers/modeling_flash_attention_utils.py

```diff
@@ -34,18 +31,129 @@
+                                e.g. for padding-free training.
+        pad_input: The function to pad inputs into one sequence and returning the respective kwargs.
+        unpad_input: The function to unpad outputs based on the kwargs (from pad_input).
+    """
+    is_fa2 = is_flash_attn_2_available()
+    is_fa3 = is_flash_attn_3_available()
+    if implementation == "flash_attention_2" or (implementation is None and is_fa2 and not is_fa3):
+        from flash_attn import flash_attn_func, flash_attn_varlen_func
+        from flash_attn.bert_padding import pad_input, unpad_input
+    else:
+        pad_input, unpad_input = _pad_input, _unpad_input
+        if implementation == "flash_attention_3" or (implementation is None and is_fa3):
+            from flash_attn_interface import flash_attn_func, flash_attn_varlen_func
+        elif is_torch_npu_available():
+            from .integrations.npu_flash_attention import npu_flash_attn_func as flash_attn_func
+            from .integrations.npu_flash_attention import npu_flash_attn_varlen_func as flash_attn_varlen_func
+        # Kernels fallback
+        else:
+            flash_attn_func = getattr(implementation, "flash_attn_func", None)
+            flash_attn_varlen_func = getattr(implementation, "flash_attn_varlen_func", None)
+            if flash_attn_varlen_func is None or flash_attn_func is None:
+                raise ValueError(
+                    f"Could not find the currently requested flash attention implementation at `{implementation}`."
+                    f"Make sure that you request a valid kernel from the hub, e.g. `kernels-community/flash-attn`."
+                )
+
+    return flash_attn_func, flash_attn_varlen_func, pad_input, unpad_input
+
+
+def lazy_import_flash_attention(implementation: Optional[str]):
+    """
+    Lazy loading flash attention and returning the respective functions + flags back
 
-def _fa3_unpad_input(hidden_states, attention_mask, unused_mask=None):
+    NOTE: For fullgraph, this needs to be called before compile while no fullgraph can
+          can work without preloading. See `_check_and_adjust_attn_implementation` in `modeling_utils`.
     """
-    FA3-compatible unpad_input function.
+    global _flash_fn, _flash_varlen_fn, _pad_fn, _unpad_fn
+    if any(k is None for k in [_flash_fn, _flash_varlen_fn, _pad_fn, _unpad_fn]):
+        _flash_fn, _flash_varlen_fn, _pad_fn, _unpad_fn = _lazy_imports(implementation)
+
+    # Depending on version and kernel some features are not supported
+    global _supports_dropout, _supports_sliding_window, _supports_determinism, _supports_softcap, _supports_s_aux
+    if any(
+        k is None
+        for k in [
+            _supports_dropout,
+            _supports_sliding_window,
+            _supports_determinism,
+            _supports_softcap,
+            _supports_s_aux,
+        ]
+    ):
+        fn_parameters = inspect.signature(_flash_varlen_fn).parameters
```

Comment:
tried to avoid inspect, but if we use it, we should just take all `kwargs` and pass them to the function if they are in the signature's parameter